### ISSUE TITLE
How do I typehint the registry pattern as suggested by the documentation?

### ISSUE THREAD

**Issue by EdwardDrapkin** [react-navigation/react-navigation#3842]
The documentation suggests creating a navigation registry here: https://reactnavigation.org/docs/navigating-without-navigation-prop.html

I'm using the officially provided Flow type definition.

With the reference capture specifically, what type am I supposed to be using?

```
const Nav = StackNavigator({}, {});

class FooComponent extends React.Component<{}> {
    navigatorRef: WhatTypeAmI = null;
    capture = (ref: WhatTypeAmI) => this.navigatorRef = ref;

    render() {
        return <Nav ref={this.capture}/>;
    }

    blah(action: NavigationAction) {
        this.navigatorRef.dispatch(action);
    }
}
```

What type should `WhatTypeAmI` be?  I looked in the flowtype file and there doesn't seem to be any concrete types that actually implement a dispatch() method.  The only things that match `dispatch` as a keyword are `NavigationProp` and `NavigationScreenProp`, which are certainly not any kind of component type.

Is this type just missing?  Looking at the code, `StackNavigator` returns type `NavigationContainer`, so should dispatch and friends be on `NavigationContainer`?

**Comment by brentvatne** [comment]
https://github.com/react-navigation/react-navigation/blob/master/flow/react-navigation.js#L458-L461

**Comment by Ashoat** [comment]
`NavigationContainer` is the type you want, but it doesn't include `dispatch` at the moment. Try this patch on your libdef file:
```diff
diff --git a/flow/react-navigation.js b/flow/react-navigation.js
index e46900c..de95773 100644
--- a/flow/react-navigation.js
+++ b/flow/react-navigation.js
@@ -549,6 +549,7 @@ declare module 'react-navigation' {
   > = React$ComponentType<NavigationContainerProps<State, Options> & Props> & {
     router: NavigationRouter<State, Options>,
     navigationOptions?: ?NavigationScreenConfig<Options>,
+    dispatch: (action: NavigationAction) => bool,
   };

   declare export type NavigationContainerProps<S: {}, O: {}> = $Shape<{
```

**Comment by EdwardDrapkin** [comment]
```
Cannot call Navigators.current.dispatch because:
 • Either property dispatch is missing in statics of
   React.Component [1].
 • Or property dispatch is missing in object type [2].

     src/screens/SignUp/components/ApplyStepsBanner.js
     117│
     118│         nav = [
     119│             () => Navigators.current && Navigators.current.dispatch(NavigationActions.popToTop()),
     120│             () => Navigators.current && Navigators.current.dispatch(NavigationActions.back()),
     121│         ];
     122│
     123│         calculateFlex() {

     flow-typed/npm/react-navigation_v1.x.x.js
 [1] 542│         Class<& React$Component<NavigationContainerProps<State, Options> & Props, any>
     543│             & { dispatch: (action: NavigationAction) => bool }>
 [2] 544│         & { router: NavigationRouter<State, Options>, navigationOptions?: ?NavigationScreenConfig<Options> };


```

Still no dice.

**Comment by EdwardDrapkin** [comment]
Since updating react-native to 0.55 (and flow to 0.67.0) I'm getting this now:

```
$ flow
Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ src/screens/FlexMobile.js:280:33

Cannot create RootNavigator element because object type [1] is not a React component.

     src/screens/FlexMobile.js
     277│                     <TokenProvider value={this.storage.token}>
     278│                         <CustomerProvider value={this.state.customerStatus}>
     279│                             <StyleProvider style={theme}>
     280│                                 <RootNavigator ref={this.captureNav}/>
     281│                             </StyleProvider>
     282│                         </CustomerProvider>
     283│                     </TokenProvider>

     flow-typed/npm/react-navigation_v1.x.x.js
 [1] 541│         Props: {}> = React$ComponentType<NavigationContainerProps<State, Options> & Props> & {
     542│         router: NavigationRouter<State, Options>,
     543│         navigationOptions?: ?NavigationScreenConfig<Options>,
     544│     };



Found 1 error
```

Does anyone working on this project actually even use Flow?

It's starting to look more and more like the flow definition needs to be scrapped and re-written from scratch, which is kind of a massive time sink, so is there anyone here who's using this library in production with the flow definition file as-is?  I can't imagine that I'm the only one running into these issues...

**Comment by Ashoat** [comment]
I use React Navigation with Flow. However, I am a volunteer with only limited time to devote to this project. I haven't yet updated to React Native 0.55, so I haven't experienced the error you're describing.

Pull requests are always welcome. (That said, I don't think scrapping the libdef makes sense... if you're about to embark on that sort of undertaking, I would appreciate some explanation of the issues you're seeing and why you think rewriting is the only solution.)

Reading through the Flow error you pasted, it looks like the latest version of Flow might've gotten a bit too strict about what component types it lets you use to declare React elements. It should be possible to use a React component that also has class-level (ie. Java "static") properties. As a starting point, I'd investigate if anybody else has been experiencing issues like this with recent versions of Flow, and if there are any existing workarounds. If there aren't, it's probably worth posting an issue to the Flow GitHub.

**Comment by Ashoat** [comment]
@EdwardDrapkin, I think I figured out the issue you were encountering in #4230. You'll still need [the patch from my earlier comment](https://github.com/react-navigation/react-navigation/issues/3842#issuecomment-379601676). If the patch with #4320 fix your issue, please let me know and we can try and get the patch committed.

**Comment by EdwardDrapkin** [comment]
We actually had enough issues with this library (incomplete or inaccurate documentation, flow support, overcomplicated API causing junior developers issues) that when we ran into an issue that would require doing a re-architecture of a feature what we wanted to implement in order to support the navigator, we decided to switch to react-native-navigation.  I'm not sure if what we were doing was just the wrong way of implementing the library, but there isn't an obviously correct way of doing anything from the documentation.  

Sorry for making you invest time if you did it on my behalf.
